### ISSUE TITLE
The result of tabnet is not better than xgboost

### ISSUE THREAD

**Issue by smile0925** [dreamquark-ai/tabnet#262]
I I tested the code for the binary classification task, but the result of TabNet is worse than the result of XGBoost. What could cause this situation?

**Comment by eduardocarvp** [comment]
Hi @smile0925,

Thanks for trying our package. It is note guaranteed at all, however, that we can have better performances with TabNet compared to XGB on every dataset and every type of problem. Even the original article shows some dataset in which CatBoost or XGBoost have better performances, suchs as KDD Churn and Upselling.

If you give me more details on the problem you are trying to solve maybe we can help you improve TabNet on your dataset a bit?

**Comment by Optimox** [comment]
Well your issue seems a bit light to give a useful answer.

You tested the code on which dataset? What are the scores you get for TabNet and XGBoost? What parameters did you use for both?

There is no such thing as a miracle algorithm, XGBoost will do worse than a linear regression for a linear problem for example. There are absolutely no guarantees or laws telling you than TabNet will outperform XGBoost systematically.

Only sure things is that TabNet results can be improved with some tweaking and understanding the data and how to tune the parameters accordingly, however you are not giving any information to try to help you.

Please give us more details.

**Comment by Ibtastic** [comment]
Hey,
I have also found the same to be true for my dataset. I'll share some details.

My problem is in the food tech domain, given historical order data like time of the day, product ordered , distance between the customer and restaurant etc(I have ~140 of such features), I am trying to predict the time(in mins) it will take for the food to reach the customer after the order is placed. I have 261060 rows in train and 11341 in test.
With Xgboost Regressor, I am able to get MAE ~7.29, RMSE~9.66.
With Tabnet Regressor(n_d=16,n_a=16,n_steps= 6), I also tried the default model, I get MAE~8.25 and RMSE~10.72. The training loss(MSE) seems to saturate pretty quickly beyond which the model just overfits.

Any insights on how to tune the model would really help!
Also, will pretraining be useful in this case?

**Comment by Optimox** [comment]
Hello @Ibtastic,

It seems that you have many steps, maybe you could try n_steps=1, 2 and 3 first and see how it goes.

**Comment by Ibtastic** [comment]
Thanks! Reducing the number of steps , did get my MAE to 8.15. 
Anything else, I should try?

**Comment by Optimox** [comment]
It's hard to give a blind help to fine tune on a dataset.
You should try a grid search with different parameters.

My main advise would be:
- optimize the loss you want to minimize, if looking for mae then use L1Loss
- use a OneCycleScheduler to speed up convergence and check if it converges within 50, 40, 30, 20 epochs. Pick the minimum number and then try some hyperparameter tuning.
- be sure to use embeddings if you have categorical variables

Again, there is no guarantee that TabNet will outperform XGBoost. It's unreasonnable to ask this for any model.

**Comment by Ibtastic** [comment]
This is helpful. Thank you!

**Comment by newbiesitl** [comment]
Same here, on my dataset Tabnet is failed to converge to near global minimal, but simple stacked dense layer beat both Tabnet and XGBoost. 
Nothing fancy to share, tabular data in millions row scale with ~30 features, 10 categorical others are numerical
